frontend/src/components/shared/StreamlitMarkdown/StreamlitMarkdown.tsx

```diff
@@ -271,4 +304,44 @@ export function LinkWithTargetBlank(props: LinkProps): ReactElement {
   )
 }
 
+export function Heading(props: HeadingProtoProps): ReactElement {
+  const { width } = props
+  const { tag, anchor, body } = props.element
+  const isSidebar = React.useContext(IsSidebarContext)
+  // st.header can contain new lines which are just interpreted as new
+  // markdown to be rendered as such.
+  const [heading, ...rest] = body.split("\n")
+
+  return (
+    <div className="stMarkdown" style={{ width }}>
+      <HeadingWithAnchor tag={tag} anchor={anchor}>
+        <RenderedMarkdown
+          source={heading}
+          allowHTML={false}
+          // this is purely an inline string
+          overrideComponents={{
+            p: Fragment,
+            h1: Fragment,
+            h2: Fragment,
+            h3: Fragment,
+            h4: Fragment,
+            h5: Fragment,
+            h6: Fragment,
+          }}
+        />
+      </HeadingWithAnchor>
+      {rest.length > 0 && (
```

Comment:
Probably would put a comment here. Explaining that we do not restrict the user so if there is a new line, we assume they just continued markdown (as was original behavior).